optuna/_hypervolume/wfg.py

```diff
@@ -6,64 +6,57 @@
-    # Time complexity: O(N^2) where N is the number of points
-    #
-    # If X, Y, Z coordinates are permutations of 0, 1, ..., N-1 and reference_point is (N, N, N),
-    # the hypervolume is calculated as the number of voxels (x, y, z) which is dominated by
-    # at least one point. If we fix x and y, this number equals the minimum of z' over all points
-    # (x', y', z') satisfying x' <= x and y' <= y. This can be efficiently computed using
-    # cumulative minimum (`np.minimum.accumulate`).
-    # In general case in which coordinates are not permutations, we can use coordinate compression
-    # to reduce to the case where the coordinates form a permutation.
-
-    assert sorted_pareto_sols.shape[1] == 3 and reference_point.shape[0] == 3
-
+    """
+    Compute hypervolume in 3D. Time complexity is O(N^2) where N is sorted_pareto_sols.shape[0].
+    If X, Y, Z coordinates are permutations of 0, 1, ..., N-1 and reference_point is (N, N, N), the
+    hypervolume is calculated as the number of voxels (x, y, z) dominated by at least one point.
+    If we fix x and y, this number is equal to the minimum of z' over all points (x', y', z')
+    satisfying x' <= x and y' <= y. This can be efficiently computed using cumulative minimum
+    (`np.minimum.accumulate`). Non-permutation coordinates can be transformed into permutation
+    coordinates by using coordinate compression.
+    """
+    assert sorted_pareto_sols.shape[1] == reference_point.shape[0] == 3
     n = sorted_pareto_sols.shape[0]
-
-    # Since sorted_pareto_sols[:, 0] is sorted, we don't have to explicitly
-    # compress coordinates for this axis
-    x_vals = np.concatenate([sorted_pareto_sols[:, 0], reference_point[:1]])
-    y_ind, y_vals = _compress_coordinate(sorted_pareto_sols[:, 1])
-    y_vals = np.concatenate([y_vals, reference_point[1:2]])
-
-    z_max = reference_point[2]
-    zs = np.full((n, n), z_max)
-    for i in range(n):
-        zs[i, y_ind[i]] = sorted_pareto_sols[i, 2]
-
-    zs = np.minimum.accumulate(zs, axis=0)
-    zs = np.minimum.accumulate(zs, axis=1)
-
-    x_delta = x_vals[1:] - x_vals[:-1]
-    y_delta = y_vals[1:] - y_vals[:-1]
-
-    return np.sum((z_max - zs) * x_delta[:, np.newaxis] * y_delta[np.newaxis, :], axis=(0, 1))
+    y_indices, y_vals = _compress_coordinate(sorted_pareto_sols[:, 1])
+    z_delta = np.zeros((n, n), dtype=float)
+    z_delta[np.arange(n), y_indices] = reference_point[2] - sorted_pareto_sols[:, 2]
+    z_delta = np.maximum.accumulate(np.maximum.accumulate(z_delta, axis=0), axis=1)
+    # The x axis is already sorted, so no need to compress this coordinate.
+    x_vals = sorted_pareto_sols[:, 0]
+    x_delta = np.concatenate([x_vals[1:], reference_point[:1]]) - x_vals
+    y_delta = np.concatenate([y_vals[1:], reference_point[1:2]]) - y_vals
+    # NOTE(nabenabe): `np.vdot(A, B)` is a faster calculation of `np.sum(A * B)`.
+    return np.vdot(x_delta[:, np.newaxis] * y_delta, z_delta)
```

Comment:
```suggestion
    return np.dot(np.dot(z_delta, y_delta), x_delta)
```

Reply:
Thanks, I benchmarked and yours is the fastest!

```python
In [2]: import numpy as np
   ...: 
   ...: 
   ...: a = np.random.random(30)
   ...: b = np.random.random(30)
   ...: C = np.random.random((30, 30))
   ...: %timeit np.vdot(a[:, None] * b, C)
   ...: %timeit np.dot(C @ b, a)
   ...: %timeit np.dot(np.dot(C, b), a)    
1.93 μs ± 39.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
1.1 μs ± 5.09 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
801 ns ± 1.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```